FogCreek/Glitch-Community
# Like story: Clarify error states on optimistic input fields

## Links
* https://like-story.glitch.me
* https://glitch.manuscript.com/f/cases/3327685/

## GIF/Screenshots:
![](http://g.recordit.co/8IjIdGPp3u.gif)

## Changes:
* when we hit an error show a default error message to the user with the truck (this just seemed like a good idea to me, but if we think it shouldn't work that way let me know, happy to change it! for example I do wonder if the truck should be reserved for user-errors and not show for things like server errors?)
* when we hit an error and then we blur the field, we hide the error and instead show the last saved good state so it's clear to the user what's there.
* some overall renaming as I found stuff a bit confusing here and there and am attempting to make it a little easier to read, though it's possible I made it even more confusing lol, please advise if you have feelings on it. I renamed useOptimisticText for example to usePassivelyTrimmedInput.

## How To Test:
* you can play around with these fields on storybook for now as shown in the gif, or if you're feeling into it you can go through all the places we use optimistic inputs, I believe I've identified all of them here:

- [x] Optimistic-markdown-input
    - [x] auth-description
        - [x] note
        - [x] collection page ( collection description)
        - [x] project page (project description)
        - [x] team page
        - [x] user page
- [x] Optimistic-text-input
    - [x] project-domain-input
    - [x] team-fields
    - [x] user-login-input
    - [x] user-name-input
- [x] Optimistic-wrapping-text-input
    - [x] collection-name-input

To test error states I added things like this:
```
  async updateNote({ note, projectId }) {
    note = _.trim(note);
    throw new Error("test error")
    await this.props.api.patch(`collections/${this.state.id}/project/${projectId}`, { annotation: note });
    this.updateProject({ note, isAddingANewNote: true }, projectId);
  }
```
but I imagine you could also toggle to offline or do something else to force an error

To compare with what we have on prod you may wish to play around with: https://long-museum.glitch.me
I added some alternating error/not error behavior to updating notes which you can see here: https://glitch.com/edit/#!/long-museum?path=src/state/collection.js:82:4
you'll notice that it doesn't show errors that don't have the exact format we expect from our api.

## Feedback I'm looking for:
- bugs! I found this to be trickier than I originally anticipated, very possible I missed something
- code style, again it's possible I'm doing something weird here. Originally we had a custom hook which inside of it called a custom hook. Now I have component that calls 2 hooks. I had originally thought to make it a component that called 3 hooks and transformed state along the way but that proved...hard? so now it's 2 hooks. Perhaps I should go back to the nesting we had before?
- ux: does this feel right to you? I *think* this is an improvement over what we have but I wouldn't be surprised if it jogs new ideas for how this should work so it's as seamless as possible. it's also difficult for me to tell because I've been looking at it for too long
- I added a default error message, does it seem clear/appropriate/on brand?
